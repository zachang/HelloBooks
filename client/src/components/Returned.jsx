import React from 'react';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Pagination } from 'react-materialize';
import { viewUserReturnAction } from '../actions/bookAction';
import UserReturn from './return/UserReturn.jsx';
import { decodeToken } from '../utils/helpers';

/**
 * Returned class declaration
 * @class Returned
 * @extends {React.Component}
 */
export class Returned extends React.Component {
  /**
   * class constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      errors: null,
      books: [],
      pageCount: null,
      limit: 15,
    };
  }

  /**
   * @method componentWillMount
   * @return {void} void
   */
  componentWillMount() {
    const userDetails = decodeToken(window.sessionStorage.getItem('token'));
    this.props.viewUserReturnAction(userDetails.id, this.state.limit, 0);
  }

  /**
   * @method componentWillReceiveProps
   * @param {object} nextProps - nextProps
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.bookState.success === false) {
      this.setState({ errors: nextProps.bookState.errors });
    }
    if (this.state.pageCount !== nextProps.bookState.pageCount) {
      this.setState({ pageCount: nextProps.bookState.pageCount });
    }
  }

  /**
   * Renders Returned component
   * @return {XML} JSX
   */
  render() {
    const userId = decodeToken(window.sessionStorage.getItem('token'));
    return (
      <div className='row'>
        <div className='section'>
          <h4 style={{ marginTop: '7%' }}>Returned Books</h4>
        </div>
        <div className='divider'
          style={{ marginTop: '-2%', marginBottom: '3%' }}
        >
        </div>

        <div className='row'>
          { (this.props.bookState.returnings) ?
            this.props.bookState.returnings.map((returning, i) =>
              <UserReturn
                key={i}
                returning={returning}
              />) : null }
        </div>

        <div className='row'>
          {
            ((this.state.pageCount) ?
              <Pagination
                items={this.state.pageCount}
                onSelect={(page) => {
                  const offset = (page - 1) * this.state.limit;
                  this.props.viewUserReturnAction(userId.id, this.state.limit,
                    offset);
                }} /> : '')
          }
        </div>
      </div>
    );
  }
}

Returned.propTypes = {
  bookState: PropTypes.object.isRequired,
  viewUserReturnAction: PropTypes.func.isRequired,
};
const mapStateToProps = state => ({
  bookState: state.bookReducer,
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ viewUserReturnAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Returned);