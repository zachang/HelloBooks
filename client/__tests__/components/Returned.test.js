import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { Returned } from '../../src/components/Returned.jsx';


describe('<Returned/>', () => {
  // create a spy function for viewUserReturnAction
  const viewUserReturnAction = sinon.spy();
  // spy on componentWillReceiveProps of Profile Page
  sinon.spy(Returned.prototype, 'componentWillMount');
  // spy on componentWillReceiveProps of Profile Page
  sinon.spy(Returned.prototype, 'componentWillReceiveProps');

  const props = {
    bookState: {
      success: true,
      errors: null,
      returnings: ['mmmmmm'],
      pageCount: null
    },
    viewUserReturnAction
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<Returned {...props} />);
    shallowComponent.setState({
      pageCount: 1,
      returnings: ['mmm'],
    });
  });

  it('renders <Returned /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.pageCount = 1;
    props.bookState.pageCount = 2;
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Returned.prototype.componentWillReceiveProps.called)
      .toEqual(true);
  });

  it('calls componentWillReceiveProps to view returned books', () => {
    props.bookState.success = false;
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Returned.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it(`should check if the 'Pagination' is defined`, () => {
    expect(shallowComponent.find('Pagination').length).toBe(1);
    shallowComponent.find('Pagination').simulate('select');
  });
});