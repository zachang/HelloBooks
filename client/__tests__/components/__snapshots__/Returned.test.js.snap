// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Returned/> renders <Returned /> component 1`] = `
<div
  className="row"
>
  <div
    className="section"
  >
    <h4
      style={
        Object {
          "marginTop": "7%",
        }
      }
    >
      Returned Books
    </h4>
  </div>
  <div
    className="divider"
    style={
      Object {
        "marginBottom": "3%",
        "marginTop": "-2%",
      }
    }
  />
  <div
    className="row"
  >
    <UserReturn
      key="0"
      returning="mmmmmm"
    />
  </div>
  <div
    className="row"
  >
    <Pagination
      activePage={1}
      items={1}
      onSelect={[Function]}
    />
  </div>
</div>
`;
